From patchwork Fri Sep 28 16:41:53 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Peng Hao <penghao122@sina.com.cn>
X-Patchwork-Id: 10620181
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 016EE15E8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 28 Sep 2018 16:47:32 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id E86FE2B7D9
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 28 Sep 2018 16:47:32 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id DC4BC2BB52; Fri, 28 Sep 2018 16:47:32 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 211B92B7D9
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 28 Sep 2018 16:47:32 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728972AbeI1XMG (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 28 Sep 2018 19:12:06 -0400
Received: from mail78-58.sinamail.sina.com.cn ([219.142.78.58]:53270 "HELO
        mail78-58.sinamail.sina.com.cn" rhost-flags-OK-OK-OK-OK)
        by vger.kernel.org with SMTP id S1728451AbeI1XMG (ORCPT
        <rfc822;kvm@vger.kernel.org>); Fri, 28 Sep 2018 19:12:06 -0400
X-Greylist: delayed 324 seconds by postgrey-1.27 at vger.kernel.org;
 Fri, 28 Sep 2018 19:12:05 EDT
Received: from unknown (HELO
 localhost.localdomain.localdomain)([110.184.35.163])
        by sina.com with ESMTP
        id 5BAE59D300005D8A; Fri, 29 Sep 2018 00:42:04 +0800 (CST)
X-Sender: penghao122@sina.com.cn
X-Auth-ID: penghao122@sina.com.cn
X-SMAIL-MID: 573004400706
From: Peng Hao <penghao122@sina.com.cn>
To: pbonzini@redhat.com, rkrcmar@redhat.com, tglx@linutronix.de,
        mingo@redhat.com, hpa@zytor.com
Cc: linux-kernel@vger.kernel.org, kvm@vger.kernel.org, x86@kernel.org,
        Peng Hao <peng.hao2@zte.com.cn>
Subject: [PATCH]  kvm/x86 : set meaningful return value
Date: Fri, 28 Sep 2018 12:41:53 -0400
Message-Id: <1538152913-3036-1-git-send-email-penghao122@sina.com.cn>
X-Mailer: git-send-email 1.8.3.1
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Peng Hao <peng.hao2@zte.com.cn>

kvm_irq_delivery_to_apic_fast()-->
    kvm_apic_map_get_dest_lapic()-->
        kvm_apic_disabled_lapic_found()
kvm_apic_map_get_dest_lapic return with bitmap==0 and dst[i]==NULL,
then (*r == -1) will be returned to qemu and "KVM: injection failed,
MSI lost(Operation not permitted)" will be recorded in qemu log. The
output is puzzling.

Signed-off-by: Peng Hao <peng.hao2@zte.com.cn>
---
 arch/x86/kvm/lapic.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kvm/lapic.c b/arch/x86/kvm/lapic.c
index fbb0e6d..a8896b3 100644
--- a/arch/x86/kvm/lapic.c
+++ b/arch/x86/kvm/lapic.c
@@ -944,7 +944,7 @@ bool kvm_irq_delivery_to_apic_fast(struct kvm *kvm, struct kvm_lapic *src,
 	int i;
 	bool ret;
 
-	*r = -1;
+	*r = -ENXIO;
 
 	if (irq->shorthand == APIC_DEST_SELF) {
 		*r = kvm_apic_set_irq(src->vcpu, irq, dest_map);
